File: go/house/house.go
package house

import "strings"

const testVersion = 1

var verse = []string{

	"the house that Jack built.",
	`the malt
that lay in`,
	`the rat
that ate`,
	`the cat
that killed`,
	`the dog
that worried`,
	`the cow with the crumpled horn
that tossed`,
	`the maiden all forlorn
that milked`,
	`the man all tattered and torn
that kissed`,
	`the priest all shaven and shorn
that married`,
	`the rooster that crowed in the morn
that woke`,
	`the farmer sowing his corn
that kept`,
	`the horse and the hound and the horn
that belonged to`,
}

func Song() string {

	verses := make([]string, 0)
	for i := range verse {
		verses = append(verses, Verse(i+1))
	}
	return strings.Join(verses, "\n\n")

}

func recursiveVerse(v int) string {
	if v > 1 {
		return verse[v-1] + " " + recursiveVerse(v-1)
	}
	if v == 1 {
		return verse[v-1]
	}
	return ""
}

func Verse(v int) string {
	return "This is " + recursiveVerse(v)
}

File: haskell/house/src/House.hs
module House
  ( rhyme
  ) where

import           Data.List

rhyme :: String
rhyme = intercalate "\n" (map verse [0 .. length list - 1])

verse :: Int -> String
verse v = "This is" ++ verse' v

verse' :: Int -> String
verse' v
  | v < 0 = ""
  | otherwise = list !! v ++ verse' (v - 1)

list =
  reverse
    [ " the horse and the hound and the horn\nthat belonged to"
    , " the farmer sowing his corn\nthat kept"
    , " the rooster that crowed in the morn\nthat woke"
    , " the priest all shaven and shorn\nthat married"
    , " the man all tattered and torn\nthat kissed"
    , " the maiden all forlorn\nthat milked"
    , " the cow with the crumpled horn\nthat tossed"
    , " the dog\nthat worried"
    , " the cat\nthat killed"
    , " the rat\nthat ate"
    , " the malt\nthat lay in"
    , " the house that Jack built.\n"
    ]
